marconi-chain-index/src/Marconi/ChainIndex/Experimental/Runner.hs

```diff
@@ -56,27 +62,81 @@ runIndexer
   -> indexer (WithDistance BlockEvent)
   -> IO ()
 runIndexer socketPath networkId _startingPoint traceName indexer = do
-  securityParam <- Utils.toException $ Utils.querySecurityParam @Void networkId socketPath
-  eventQueue <- STM.newTBQueueIO $ fromIntegral securityParam
-  cBox <- Concurrent.newMVar indexer
   c <- Tracing.defaultConfigStdout
-  Async.concurrently_
-    ( withTrace c traceName $ \trace ->
-        let io =
-              withChainSyncEventEpochNoStream
-                socketPath
-                networkId
-                [Core.genesis]
-                (mkEventStream eventQueue . chainSyncEventStreamLogging trace)
-            handleException NoIntersectionFound =
-              Trace.logError trace $
-                Pretty.renderStrict $
-                  Pretty.layoutPretty
-                    Pretty.defaultLayoutOptions
-                    "No intersection found"
-         in io `catch` handleException
-    )
-    (readEvent eventQueue cBox)
+  withTrace c traceName $ \trace -> runChainSyncClientWithRetries trace
+  where
+    secondsToWait :: Int
+    secondsToWait = 30
+
+    runChainSyncClientWithRetries
+      :: Trace IO Text
+      -> IO ()
+    runChainSyncClientWithRetries trace = do
```

Comment:
A classical strategy for retry is to start with a quick retry and to space them more and more. We can either implement such strategy now or do it in another ticket
We could use something like:

```
data RetryState =
  RetryState
  { secondsBeforeNextRetry :: Word
  , initialWait :: Word
  , maxWait :: Word
  }
```

We pass this type to `runChainSyncClientWithRetries` (or we do use a state monad), erinitialise to `initialWait` on success, double the time on each retry.

Reply:
Doesn't seem too complicated. I can do it as part of this PR.

Reply:
The change is done. However, if the node connection is re-established, then we increment the number of seconds before retry based on the previous number. Not sure how to make this better.

Here's the new log:

```
[marconi-chain-index:Warning:11] [2023-08-14 17:02:37.15 UTC] Could not connect to Cardano node: socket file /home/kolam/cardano-node/preprod/cardano-node.socket does not exist. Retrying in 30s ...
[marconi-chain-index:Warning:11] [2023-08-14 17:03:07.18 UTC] Could not connect to Cardano node: socket file /home/kolam/cardano-node/preprod/cardano-node.socket does not exist. Retrying in 60s ...
[marconi-chain-index:Info:11] [2023-08-14 17:04:07.33 UTC] Resumable points for each indexer:
    [ ChainPoint(Slot 993935, BlockHash ffb2e3f74984692541a11301c8e5851be8db42fcd65230f049814824c700bd20)
    , ChainPoint(Slot 993935, BlockHash ffb2e3f74984692541a11301c8e5851be8db42fcd65230f049814824c700bd20)
    , ChainPoint(Slot 993935, BlockHash ffb2e3f74984692541a11301c8e5851be8db42fcd65230f049814824c700bd20)
    , ChainPoint(Slot 993935, BlockHash ffb2e3f74984692541a11301c8e5851be8db42fcd65230f049814824c700bd20) ]
[marconi-chain-index:Info:11] [2023-08-14 17:04:07.34 UTC] Starting from ChainPoint(Slot 993935, BlockHash ffb2e3f74984692541a11301c8e5851be8db42fcd65230f049814824c700bd20). 
[marconi-chain-index:Info:11] [2023-08-14 17:04:17.34 UTC] Synchronising (4.44%). Current synced point is ChainPoint(Slot 1614855, BlockHash 770d8355c9958b9cebc210e95c6d209bec09020788f98b87f7c3ecd9cde10e41) and current node tip is ChainTip(Slot 36349430, BlockHash 83cd3e7e437bc7ea9a9446455356786b158934c55ef90fe
b3fdd57a6fb6cee6c, BlockNo 1270014). Processed 31103 blocks and 1 rollbacks in the last 10s (3110 blocks/s).
[marconi-chain-index:Info:11] [2023-08-14 17:04:27.35 UTC] Synchronising (5.95%). Current synced point is ChainPoint(Slot 2162421, BlockHash 2f7107e6d00f4a4867b8525d31150f45180c033b285c93564a150e6974756cdb) and current node tip is ChainTip(Slot 36349430, BlockHash 83cd3e7e437bc7ea9a9446455356786b158934c55ef90fe
b3fdd57a6fb6cee6c, BlockNo 1270014). Processed 27247 blocks and 0 rollbacks in the last 10s (2721 blocks/s).
[marconi-chain-index:Info:11] [2023-08-14 17:04:37.35 UTC] Synchronising (7.44%). Current synced point is ChainPoint(Slot 2703595, BlockHash 582dda8c9997158e94a631c5d395c971c448580319c45f3bbb58f7d04d8e0fe8) and current node tip is ChainTip(Slot 36349430, BlockHash 83cd3e7e437bc7ea9a9446455356786b158934c55ef90fe
b3fdd57a6fb6cee6c, BlockNo 1270014). Processed 26845 blocks and 0 rollbacks in the last 10s (2684 blocks/s).
[marconi-chain-index:Info:11] [2023-08-14 17:04:39.35 UTC] Stopping indexing. Waiting for indexers to finish their work (timeout after 180s) ...
[marconi-chain-index:Info:11] [2023-08-14 17:04:39.35 UTC] Done!
[marconi-chain-index:Warning:11] [2023-08-14 17:04:39.35 UTC] Could not connect to Cardano node: socket file /home/kolam/cardano-node/preprod/cardano-node.socket does not exist. Retrying in 120s ...
```

See the last retry log.

Reply:
We could probably introduce a `baseTime` and revert to it on success. I'm more a afraid of a potential memory leak with the current version, as I'm not sure the recursive calls to retry can be optimised.

Reply:
> We could probably introduce a baseTime and revert to it on success.

The problem is that we don't know what "on success" is. The function just runs indefinitely. Unless we consider that the function is "successful" if it has run after a certain amount of time, but that seems arbitrary.

> I'm more a afraid of a potential memory leak with the current version, as I'm not sure the recursive calls to retry can be optimised.

Good point. Do you have any enhancement ideas?

Reply:
Sorry, I didn't take the time to think about it. So no, nothing at the moment. We can look at it when I'll be back or you can monitor it and check if it has a memory impact. In the current state, an infinite and linear retry time might be easier/more stable, but I let you decide.